tests/unit/attestations/test_services.py

```diff
@@ -0,0 +1,421 @@
+        "publisher_name",
+        [
+            "github",
+            "gitlab",
+        ],
+    )
+    def test_generate_provenance_succeeds(self, publisher_name: str, monkeypatch):
+        integrity_service = IntegrityService(
+            storage=pretend.stub(),
+            metrics=pretend.stub(),
+        )
+
+        if publisher_name == "github":
+            publisher = GitHubPublisher(
+                repository="fake-repository",
+                workflow="fake-workflow",
+            )
+        else:
+            publisher = GitLabPublisher(
+                repository="fake-repository",
+                environment="fake-env",
+            )
+
+        monkeypatch.setattr(
+            services,
+            "_publisher_from_oidc_publisher",
+            lambda s: publisher,
+        )
+
+        provenance = integrity_service.generate_provenance(
+            pretend.stub(),
+            [VALID_ATTESTATION],
+        )
+
+        assert provenance == Provenance(
+            attestation_bundles=[
+                AttestationBundle(
+                    publisher=publisher,
+                    attestations=[VALID_ATTESTATION],
+                )
+            ]
+        )
+
+    def test_persist_provenance_succeeds(self, db_request):
+        provenance = Provenance(
+            attestation_bundles=[
+                AttestationBundle(
+                    publisher=GitHubPublisher(
+                        repository="fake-repository",
+                        workflow="fake-workflow",
+                    ),
+                    attestations=[VALID_ATTESTATION],
+                )
+            ]
+        )
+
+        @pretend.call_recorder
+        def storage_service_store(path, file_path, *_args, **_kwargs):
+            expected = provenance.model_dump_json().encode("utf-8")
+            with open(file_path, "rb") as fp:
+                assert fp.read() == expected
+
+            assert path.suffix == ".provenance"
+
+        integrity_service = IntegrityService(
+            storage=pretend.stub(store=storage_service_store),
+            metrics=pretend.stub(),
+        )
+        assert (
+            integrity_service.persist_provenance(provenance, FileFactory.create())
+            is None
+        )
+
+    def test_get_provenance_digest(self, db_request):
+        file = FileFactory.create()
+        AttestationFactory.create(file=file)
+        FileEventFactory.create(
+            source=file,
+            tag=EventTag.File.FileAdd,
+            additional={"publisher_url": "fake-publisher-url"},
+        )
+
+        with tempfile.NamedTemporaryFile() as f:
+            integrity_service = IntegrityService(
+                storage=pretend.stub(get=pretend.call_recorder(lambda p: f)),
+                metrics=pretend.stub(),
+            )
+
+            assert (
+                integrity_service.get_provenance_digest(file)
+                == hashlib.file_digest(f, "sha256").hexdigest()
+            )
```

Comment:
Flagging: I think we need to not mock at this layer to gain more confidence here -- `db_request` should now have a proper `IntegrityService` injected into it that we can use to create, persist, and digest an actual attestation object.

Reply:
Done -- I've added a non-default `storage_service` fixture that can be injected for these kinds of tests/